### PR TITLE
Windows alternate user support for execute resources

### PR DESCRIPTION
**Pull request by adamedx**
### Windows alternate user identity execute support

The `execute` resource and simliar resources such as `script`, `batch`, and `powershell_script`
now support the specification of credentials on Windows so that the resulting process
is created with the security identity that corresponds to those credentials. This gives these resources on Windows the same capability they have on *nix systems to run processes as an alternate identity.
#### Properties

The following properties are new or updated for the `execute`, `script`, `batch`, and
`powershell_script` resources and any resources derived from them:
-   `user`</br>
  **Ruby types:** String</br>
  The user name of the user identity with which to launch the new process.
  Default value: `nil`. The user name may optionally be specifed
  with a domain, i.e. `domain\user` or `user@my.dns.domain.com` via Universal Principal Name (UPN)
  format. It can also be specified without a domain simply as `user` if the domain is
  instead specified using the `domain` attribute. On Windows only, if this property is specified, the `password`
  property **must** be specified.
-   `password`</br>
  **Ruby types** String</br>
  _Windows only:_ The password of the user specified by the `user` property.
  Default value: `nil`. This property is mandatory if `user` is specified on Windows and may only
  be specified if `user` is specified. The `sensitive` property for this resource will
  automatically be set to `true` if `password` is specified.
-   `domain`</br>
  **Ruby types** String</br>
  _Windows only:_ The domain of the user user specified by the `user` property.
  Default value: `nil`. If not specified, the user name and password specified
  by the `user` and `password` properties will be used to resolve
  that user against the domain in which the system running Chef client
  is joined, or if that system is not joined to a domain it will resolve the user
  as a local account on that system. An alternative way to specify the domain is to leave
  this property unspecified and specify the domain as part of the `user` property.

### CODE REVIEW

File: lib/chef/mixin/user_identity.rb [hunk 1]
```diff
@@ -0,0 +1,81 @@
+#
+# Author:: Adam Edwards (<adamed@chef.io>)
+# Copyright:: Copyright (c) 2015 Chef Software, Inc.
```
Comment (tyler-ball) [inline lib/chef/mixin/user_identity.rb]: 2016

File: lib/chef/provider/execute.rb [hunk 2]
```diff
@@ -19,15 +19,22 @@
...
       extend Forwardable
 
+      include Chef::Mixin::UserIdentity
+
       provides :execute
 
-      def_delegators :@new_resource, :command, :returns, :environment, :user, :group, :cwd, :umask, :creates
+      def_delegators :@new_resource, :command, :returns, :environment, :user, :domain, :password, :group, :cwd, :umask, :creates
+
+      def initialize(new_resource, run_context)
```
Comment (tyler-ball) [inline lib/chef/provider/execute.rb]: This is unnecessary if all it does is call `super`

File: lib/chef/provider/script.rb [hunk 3]
```diff
@@ -69,7 +70,38 @@ def set_owner_and_group
...
-        FileUtils.chown(new_resource.user, new_resource.group, script_file.path)
+        if ! Chef::Platform.windows?
+          FileUtils.chown(new_resource.user, new_resource.group, script_file.path)
+        else
+          grant_alternate_user_read_access
+        end
+      end
+
+      def grant_alternate_user_read_access
+        return if new_resource.user.nil?
```
Comment (tyler-ball) [inline lib/chef/provider/script.rb]: So in a Windows world, if `user` is nil it will just keep the same owner as whoever create it, correct?

Comment (adamedx) [inline lib/chef/provider/script.rb]: Yes, that's correct (and that's what is currently happening in Chef + Windows today for script resources).

File: lib/chef/mixin/user_identity.rb [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+          end
+        end
+
+        { domain: domain, user: user }
+      end
+
+      protected(:validate_identity)
+      protected(:validate_identity_platform)
+      protected(:validate_identity_syntax)
+      protected(:qualify_user)
```
Comment (lamont-granquist) [inline lib/chef/mixin/user_identity.rb]: `protected` is super-weird in ruby, you probably mean `private` here (`private` methods are still accessible from subclasses).

Comment (lamont-granquist) [inline lib/chef/mixin/user_identity.rb]: http://stackoverflow.com/questions/3534449/why-does-ruby-have-both-private-and-protected-methods

"It is important to note that this is different from the way Java or C++ works. private in Ruby is similar to protected in Java/C++ in that subclasses have access to the method. In Ruby, there is no way to restrict access to a method from its subclasses like you can with private in Java."

File: lib/chef/provider/execute.rb [hunk 5]
```diff
@@ -52,6 +55,7 @@ def timeout
       end
 
       def action_run
+        validate_identity(new_resource.user, new_resource.password, new_resource.domain)
```
Comment (lamont-granquist) [inline lib/chef/provider/execute.rb]: should this go into define_resource_requirements instead?

File: lib/chef/mixin/user_identity.rb (hunk 4 above)
Comment (adamedx) [inline lib/chef/mixin/user_identity.rb]: Got it, will switch to `private` @lamont-granquist.

**Comment by btm** [comment]
Replaced by https://github.com/chef/chef/pull/5764

**Pull request closed without merging**